malsim/agents/searchers.py

```diff
@@ -1,137 +1,98 @@
+        'randomize': False,
+        # Whether to randomize next target selection, still respecting the
+        # policy of the agent (e.g. BFS or DFS).
+        'seed': None,
+        # The random seed to initialize the randomness engine with.
+    }
 
-        self.current_target = None if done else self.current_target
-        action = 0 if done else 1
-        if action == 0:
-            logger.debug(
-                "Attacker Breadth First agent does not have "
-                "any valid targets it will terminate"
-            )
+    def __init__(self, agent_config: dict) -> None:
+        """Initialize a BFS agent.
 
-        return (action, self.current_target)
+        Args:
+            agent_config: Dict with settings to override defaults
+        """
+        self.targets: deque[AttackGraphNode] = deque()
+        self.current_target: Optional[AttackGraphNode] = None
 
-    @staticmethod
-    def select_next_target(
-        current_target: int,
-        targets: Union[List[int], Deque[int]],
-        attack_surface: Set[int],
-    ) -> int:
-        # If the current target was not compromised, put it
-        # back, but on the bottom of the stack.
-        if current_target in attack_surface:
-            targets.appendleft(current_target)
-            current_target = targets.pop()
+        self.settings = self.default_settings | agent_config
 
-        while current_target not in attack_surface:
-            if len(targets) == 0:
-                return None, True
+        self.rng = np.random.default_rng(
+            self.settings['seed'] or np.random.SeedSequence()
+        )
 
-            current_target = targets.pop()
+    def get_next_action(
+        self, agent: MalSimAgentStateView, **kwargs
+    ) -> Optional[AttackGraphNode]:
+        self._update_targets(agent.action_surface)
+        self._select_next_target()
 
-        return current_target, False
+        return self.current_target
 
+    def _update_targets(self, action_surface: list[AttackGraphNode]):
 
-class DepthFirstAttacker:
-    def __init__(self, agent_config: dict) -> None:
-        self.current_target = -1
-        self.targets: List[int] = []
-        seed = (
-            agent_config["seed"]
-            if agent_config.get("seed", None)
-            else np.random.SeedSequence().entropy
-        )
-        self.rng = (
-            np.random.default_rng(seed)
-            if agent_config.get("randomize", False)
-            else None
-        )
+        # action surface does not have a guaranteed order,
+        # so for the agent to be deterministic we need to sort
+        action_surface.sort(key=lambda n: n.id)
 
-    def compute_action_from_dict(self, observation: Dict[str, Any], mask: tuple):
-        new_targets, surface_indexes = get_new_targets(observation, self.targets, mask)
+        new_targets = [
+            step
+            for step in action_surface
+            if step not in self.targets and not step.is_compromised()
```

Comment:
Would a compromised step ever end up in an agent's attack surface?

Reply:
Yes, nodes already compromised are still part of the attack surface since we do not know whether or not we want to allow them to be re-compromised or not. We could have this be a parameter for the `get_attack_surface` method, but I think just having it return a set as we were planning to do is cleaner.

Reply:
This means that the attack surface grows as the simulation proceeds. At the end of a simulation where the agent has compromised almost everything the attack surface will be a copy basically of the attack graph. I think a better interface is needed here

Reply:
That is correct and has been brought up before. The attacker also tracks its reached attack steps which also end up covering the entire graph towards the end of simulations.

Reply:
In #98, I introduced a new variable `new_attack_surface` that keeps track of only the new nodes that became available in the just executed step. Whether to keep already compromised nodes in the attack surface is a good idea is an open topic. I am resolving this for now.